### ISSUE TITLE
Custom ResetCredentialEmail does not work after upgrade to Keycloak 21

### ISSUE THREAD

**Issue by fvitolo** [keycloak/keycloak#19530]
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the latest release

### Area

authentication

### Describe the bug

Hello,
in our keycloak implementation we need to customise the ResetCredentialEmail class.

The goal is to pass in the success message, a string containing the type of operation performed and the address to which the email was sent, so that after submitting the forgot password form, we can read this information on the login page and view a resume of the operation before being able to return to the actual login.

To do this, we have extended the ResetCredentialEmail class and created the related Factory as follows:

```
public class CustomResetCredentialEmail extends ResetCredentialEmail {

    private static final Logger logger = Logger.getLogger(CustomResetCredentialEmail.class);
    private static final String FORGOT_PASSWORD_EMAIL_SENT = "forgot-password-email-sent";
    private static final String FORGOT_PASSWORD_EMAIL_SEPARATOR = "|";

    @Override
    public void authenticate(AuthenticationFlowContext context) {
        logger.info("Custom Reset Credential Email");
        UserModel user = context.getUser();
        AuthenticationSessionModel authenticationSession = context.getAuthenticationSession();
        String username = authenticationSession.getAuthNote(AbstractUsernameFormAuthenticator.ATTEMPTED_USERNAME);

        String forgotPasswordEmailMessage = String.format("%s%s%s",FORGOT_PASSWORD_EMAIL_SENT,FORGOT_PASSWORD_EMAIL_SEPARATOR,username);

        // we don't want people guessing usernames, so if there was a problem obtaining the user, the user will be null.
        // just reset login for with a success message
        if (user == null) {
            context.forkWithSuccessMessage(new FormMessage(forgotPasswordEmailMessage));
            return;
        }

        String actionTokenUserId = authenticationSession.getAuthNote(DefaultActionTokenKey.ACTION_TOKEN_USER_ID);
        if (actionTokenUserId != null && Objects.equals(user.getId(), actionTokenUserId)) {
            logger.debugf("Forget-password triggered when reauthenticating user after authentication via action token. Skipping " + PROVIDER_ID + " screen and using user '%s' ", user.getUsername());
            context.success();
            return;
        }


        EventBuilder event = context.getEvent();
        // we don't want people guessing usernames, so if there is a problem, just continuously challenge
        if (user.getEmail() == null || user.getEmail().trim().length() == 0) {
            event.user(user)
                .detail(Details.USERNAME, username)
                .error(Errors.INVALID_EMAIL);

            context.forkWithSuccessMessage(new FormMessage(forgotPasswordEmailMessage));
            return;
        }

        int validityInSecs = context.getRealm().getActionTokenGeneratedByUserLifespan(ResetCredentialsActionToken.TOKEN_TYPE);
        int absoluteExpirationInSecs = Time.currentTime() + validityInSecs;

        // We send the secret in the email in a link as a query param.
        String authSessionEncodedId = AuthenticationSessionCompoundId.fromAuthSession(authenticationSession).getEncodedId();
        ResetCredentialsActionToken token = new ResetCredentialsActionToken(user.getId(), user.getEmail(), absoluteExpirationInSecs, authSessionEncodedId, authenticationSession.getClient().getClientId());
        String link = UriBuilder
            .fromUri(context.getActionTokenUrl(token.serialize(context.getSession(), context.getRealm(), context.getUriInfo())))
            .build()
            .toString();
        long expirationInMinutes = TimeUnit.SECONDS.toMinutes(validityInSecs);
        try {
            context.getSession()
                .getProvider(EmailTemplateProvider.class)
                .setRealm(context.getRealm())
                .setUser(user)
                .setAuthenticationSession(authenticationSession)
                .sendPasswordReset(link, expirationInMinutes);

            event.clone().event(EventType.SEND_RESET_PASSWORD)
                .user(user)
                .detail(Details.USERNAME, username)
                .detail(Details.EMAIL, user.getEmail()).detail(Details.CODE_ID, authenticationSession.getParentSession().getId()).success();
            context.forkWithSuccessMessage(new FormMessage(forgotPasswordEmailMessage));
        } catch (EmailException e) {
            event.clone().event(EventType.SEND_RESET_PASSWORD)
                .detail(Details.USERNAME, username)
                .user(user)
                .error(Errors.EMAIL_SEND_FAILED);
            ServicesLogger.LOGGER.failedToSendPwdResetEmail(e);
            Response challenge = context.form()
                .setError(Messages.EMAIL_SENT_ERROR)
                .createErrorPage(Response.Status.INTERNAL_SERVER_ERROR);
            context.failure(AuthenticationFlowError.INTERNAL_ERROR, challenge);
        }
    }

}
```

```
public class CustomResetCredentialEmailFactory implements AuthenticatorFactory, ConfigurableAuthenticatorFactory {

    public static final String PROVIDER_ID = "reset-credential-email";
    private static final CustomResetCredentialEmail SINGLETON = new CustomResetCredentialEmail();

    @Override
    public String getDisplayType() {
        return "Send Reset Email";
    }

    @Override
    public String getReferenceCategory() {
        return null;
    }

    @Override
    public boolean isConfigurable() {
        return false;
    }

    public static final AuthenticationExecutionModel.Requirement[] REQUIREMENT_CHOICES = {
        AuthenticationExecutionModel.Requirement.REQUIRED
    };

    @Override
    public AuthenticationExecutionModel.Requirement[] getRequirementChoices() {
        return REQUIREMENT_CHOICES;
    }

    @Override
    public boolean isUserSetupAllowed() {
        return false;
    }

    @Override
    public String getHelpText() {
        return "Send email to user and wait for response.";
    }

    @Override
    public List<ProviderConfigProperty> getConfigProperties() {
        return null;
    }

    @Override
    public void close() {

    }

    @Override
    public Authenticator create(KeycloakSession session) {
        return SINGLETON;
    }

    @Override
    public void init(Config.Scope config) {

    }

    @Override
    public void postInit(KeycloakSessionFactory factory) {

    }

    @Override
    public String getId() {
        return PROVIDER_ID;
    }
}
```
the classes were deployed in a jar containing also the following file, src/main/resources/META-INF/services/org.keycloak.authentication.AuthenticatorFactory with the name of the factory inside.

From the keycloak logs, it appears that the authenticator is correctly loaded.

![image](https://user-images.githubusercontent.com/103186799/229809671-7bbae427-5987-49d0-8385-10003e320a6a.png)

This implementation was working up to version 20 of keycloak.

After the upgrade to keycloak 21, it appears that the authenticate method of the CustomResetCredentialEmail class is no longer invoked.

### Version

21.0.1

### Expected behavior

After successfully executing the forgot password, the context contains a message stating the address to which the reset mail was sent.

![image](https://user-images.githubusercontent.com/103186799/230050295-c57f39b5-b117-4941-8731-0d316d1333ab.png)

### Actual behavior

It appears that in the forgot password flow, the authenticate method of the default ResetCredentialEmail class is still invoked.

### How to Reproduce?

Include in the /opt/keycloak/providers/ directory of the keycloak deployment the jar containing the above classes, also including the file src/main/resources/META-INF/services/org.keycloak.authentication.AuthenticatorFactory.

Perform a forgot password operation from the keycloak login form.

**Comment by stianst** [comment]
Keycloak has never really supported overwriting an existing provider for an SPI. This has probably just worked in the past by chance rather than expected behaviour.

What you can do is to override the default provider used for an SPI. This can be done either by setting a higher order in the ProviderFactory, or by explicitly setting it through config (I think it's `--spi-<spi name>-default=<provider id>`.

**Comment by stianst** [comment]
@pedroigor @vmuzikar FIY

**Comment by fvitolo** [comment]
Hi @stianst , thanks for the reply.

Actually in my opinion being able to override an existing provider might be useful, because sometimes you only need to change a small behaviour and it might be inconvenient to have to recreate an entire authenticator and custom flow to do so (as in my specific case).

Also because writing an authenticator from scratch means having a more or less in-depth knowledge of the keycloak source code, which to date is difficult to acquire unless you do a lot of reverse engineering (unless I have missed a guide on the subject).

That said, could one have more detail on these points? 

> What you can do is to override the default provider used for an SPI. This can be done either by setting a higher order in the ProviderFactory, or by explicitly setting it through config (I think it's --spi-<spi name>-default=<provider id>.

Even a detailed reference in the documentation would be fine.

**Comment by mposolda** [comment]
This should be already fixed in latest Keycloak main.

@fvitolo @derkoe Do you please have opportunity to try with latest Keycloak main build? Or possibly with latest Keycloak nightly release https://github.com/keycloak/keycloak/releases/tag/nightly ?

It is still recommended to do what @stianst suggested as this worked in the past by chance and can possibly stop working anytime in the future.

**Comment by mposolda** [comment]
> Hi @mposolda , thanks for the suggestion but even with the nightly version my custom provider is not used.

This is quite strange. I will try to test with your provider in the following days...

@derkoe How about you? Can you see it fixed with latest Keycloak?

> 
> As already asked could i have more details on what @stianst suggested?
> 
> > What you can do is to override the default provider used for an SPI. This can be done either by setting a higher order in the ProviderFactory, or by explicitly setting it through config (I think it's --spi-<spi name>-default=<provider id>.

Keycloak has 2 type of providers:
- Providers, which are supposed to have only single implementation used inside Keycloak. For example `RealmProvider` is typical case as you usually have data stored only in the DB
- Providers, which are supposed to have multiple implementations. The `Authenticator SPI` falls into that category as you have many different authenticator implementations in the Keycloak runtime.

The note from Stian about default provider is mostly relevant for the 1st case, but not that for the 2nd case. So likely not very helpful for your case with custom authenticator.

**Comment by mposolda** [comment]
> Also because writing an authenticator from scratch means having a more or less in-depth knowledge of the keycloak source code, which to date is difficult to acquire unless you do a lot of reverse engineering (unless I have missed a guide on the subject).

You already wrote custom authenticator, so looks that you already have this knowledge? Besides that, we have docs in the server development guide [1] and quickstart for custom authenticator [2]. In your case, what you can do is just 2 steps:

- Change field CustomResetCredentialEmailFactory.PROVIDER_ID to use different provider_id than existing Keycloak provider `public static final String PROVIDER_ID = "custom-reset-credential-email";`. Will be probably good to override `getDisplayType()` as well and add some different text than default KEycloak factory has
- Configure different `Reset Credential Flow` (rather copy from existing flow) and then replace existing Reset-credential-email authenticator from KEycloak and replace it with yours. Then bind your new authentication flow as "Reset Credentials Flow"

More docs about flows [3]

[1] https://www.keycloak.org/docs/latest/server_development/index.html#_auth_spi
[2] https://github.com/keycloak/keycloak-quickstarts/tree/latest/action-token-authenticator
[3] https://www.keycloak.org/docs/latest/server_admin/#_authentication-flows

**Comment by fvitolo** [comment]
Thank you, I finally solved the problem as you suggested. I changed the provider id and built a new authentication flow.

The only problem I have now is to apply this authentication flow dynamically to all the realms already created, can this be done via API?

**Comment by mposolda** [comment]
@fvitolo Yes, it can be done by admin REST API.

Everything, which is possible to do with new admin console, is doable with admin REST API as well, because admin console just uses admin REST API under the covers.

I think the best is, that you somehow "decode" what are REST API requests used when you do this with the new admin console (for example with Chrome-developers or Firebug for FF, it can be seen what requests were sent by browser). Then same admin REST API requests can be "copied" for all the other realms where you want this to be used.

**Comment by fvitolo** [comment]
Hi @mposolda, that's exactly what we are doing, there are several calls to implement and the idea is to:

- clone the current reset-credential-flow
- add the new authenticator
- remove the default authenticator
- change the priority of the new atuhenticator, since once created it is at the bottom

The questions I would like to ask are:
- Is there by any chance an API that takes as input a complete json of all flow data (including executions) to replace the actions above with a single call?
- Is there an alternative way to change the priority of executions than using the /raise-priority and /lower-priority APIs n-times?

Translated with www.DeepL.com/Translator (free version)

**Comment by mposolda** [comment]
> Hi @mposolda, that's exactly what we are doing, there are several calls to implement and the idea is to:
> 
>     * clone the current reset-credential-flow
> 
>     * add the new authenticator
> 
>     * remove the default authenticator
> 
>     * change the priority of the new atuhenticator, since once created it is at the bottom
> 
> 
> The questions I would like to ask are:
> 
>     * Is there by any chance an API that takes as input a complete json of all flow data (including executions) to replace the actions above with a single call?

No, I think we don't have anything like that

> 
>     * Is there an alternative way to change the priority of executions than using the /raise-priority and /lower-priority APIs n-times?
No, I think we don't have anything like that

In shortcut: What you're already doing is likely the best way to address this for all your realms

**Comment by stianst** [comment]
Added an umbrella issue for these issues here:
https://github.com/keycloak/keycloak/issues/19867

**Comment by mposolda** [comment]
duplicates #19867

So I am closing this. This is fixed in Keycloak main and hence should be resolved in Keycloak 21.1.1 (will be probably released in next few days). Please re-test with Keycloak 21.1.1 once it is released and re-open if you still see issues.
